### PR TITLE
Update Dropout(12) forward kernel with training_mode input.

### PR DESCRIPTION
**Pull request by codemzs**
Adds `training_mode` input to Dropout(12) forward kernel which allows to switch between training and inferencing mode dynamically and brings the kernel up to date with [Dropout-12](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Dropout) spec.

### CODE REVIEW

File: orttraining/orttraining/training_ops/cuda/nn/dropout.cc [hunk 1]
```diff
@@ -52,6 +77,22 @@ Status Dropout<T1, T2>::ComputeInternal(OpKernelContext* context) const {
   //Get mask_data
   auto mask = context->Output(1, shape);
   ORT_ENFORCE(!mask || mask->Shape().Size() == N);
+
+  const Tensor* training_mode = context->Input<Tensor>(2);
+  //Check for inference mode.
+  if (!trainable_dropout && (training_mode == nullptr || *(training_mode->Data<bool>()) == false)) {
```
Comment (jessebenson) [inline orttraining/orttraining/training_ops/cuda/nn/dropout.cc:83]: TrainableDropout was using `ratio` input to decide between inference mode and training mode.  Specifically, `ratio == 0` indicated inference mode.  We seem to have lost that with these changes.  Can we keep that for backwards compatibility, until TrainableDropout is removed?

File: orttraining/orttraining/training_ops/cpu/nn/dropout_op.cc [hunk 2]
```diff
@@ -75,16 +80,20 @@ Status Dropout<T1, T2>::Compute(OpKernelContext* context) const {
     return gsl::make_span(temp_mask_buffer.get(), X_shape.Size());
   }();
 
-  ORT_ENFORCE(Y->Shape() == X_shape, "X and Y should have the same shape");
   ORT_ENFORCE(!mask || mask->Shape() == X_shape, "X and mask should have the same shape");
 
-  if (ratio_value == 0.0f) {
+  const Tensor* training_mode = context->Input<Tensor>(2);
+  if (!trainable_dropout && (training_mode == nullptr || *(training_mode->Data<bool>()) == false)) {
```
Comment (jessebenson) [inline orttraining/orttraining/training_ops/cpu/nn/dropout_op.cc:86]: Same comment here - TrainableDropout was using `ratio == 0` to decide to go the fast-path of just memcpy.  Preserve for now?

File: orttraining/orttraining/training_ops/cuda/nn/dropout.cc (hunk 1 above)
Comment (jessebenson) [reply orttraining/orttraining/training_ops/cuda/nn/dropout.cc:83]: Even if the RNG happens to do the right thing when ratio == 0 (I'm not sure) - this codepath is faster.

Comment (codemzs) [reply orttraining/orttraining/training_ops/cuda/nn/dropout.cc:83]: Fair point.

---
In reply to: [419723580](https://github.com/microsoft/onnxruntime/pull/3805#discussion_r419723580) [](ancestors = 419723580)

**Comment by codemzs** [comment]
Thanks @jessebenson for the review!

**Pull request merged**